app/scripts/metamask-controller.js

```diff
@@ -588,9 +588,14 @@ export default class MetamaskController extends EventEmitter {
       {
         messenger: nftControllerMessenger,
         chainId: this.networkController.state.providerConfig.chainId,
-        onPreferencesStateChange:
-          this.preferencesController.store.subscribe.bind(
-            this.preferencesController.store,
+        onPreferencesStateChange: (listener) =>
```

Comment:
Do we need to do this for `nftDetectionController` as well?

Reply:
I think we might, i was wondering if the detection was broken because of this too, i can test it 🙏

Reply:
I confirm that the detection is also broken because of the same problem (not getting the right address);

If i try to do same fix for DetectionController exp:

```
 onPreferencesStateChange: (listener) =>
        this.controllerMessenger.subscribe(
          `AccountsController:selectedAccountChange`,
          (newlySelectedInternalAccount) => {
            listener({
              selectedAddress: newlySelectedInternalAccount.address,
              useNftDetection:
                this.preferencesController.store.getState().useNftDetection,
            });
          },
        ), 
```

It will fix the detection but only when the user switches to another account because we are listening on this
`AccountsController:selectedAccountChange`
Which is not the behaviour we want, we want the user to enable the autoDetection in settings and once he is back to the same account he can see his detected NFTs.

Still looking into how to properly fix that and pass the correct account to the DetectionController